servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilterBuilder.java

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * A builder for {@link RetryingHttpRequesterFilter}, which puts an upper bound on retry attempts.
+ * To configure the maximum number of retry attempts see {@link #maxTotalRetries(int)}.
+ */
+public final class RetryingHttpRequesterFilterBuilder {
+    private boolean waitForLb = true;
+    private boolean ignoreSdErrors;
+
+    private int maxRetries = 3;
+
+    @Nullable
+    private Function<HttpResponseMetaData, BackOffPolicy> retryForResponsesMapper;
+
+    @Nullable
+    private BiFunction<HttpRequestMetaData, Throwable, BackOffPolicy>
+            retryForRequestsMapper = new DefaultRequestRetryPolicyBuilder().build();
+
+    /**
+     * By default, automatic retries wait for the associated {@link LoadBalancer} to be ready before triggering a
+     * retry for requests. This behavior may add latency to requests till the time the load balancer is ready
+     * instead of failing fast. This method allows controlling that behavior.
+     *
+     * @param waitForLb Whether to wait for the {@link LoadBalancer} to be ready before retrying requests.
+     * @return {@code this}.
+     */
+    public RetryingHttpRequesterFilterBuilder waitForLoadBalancer(final boolean waitForLb) {
+        this.waitForLb = waitForLb;
+        return this;
+    }
+
+    /**
+     * By default, fail a request if the last signal from the associated {@link ServiceDiscoverer} was an error.
+     * This method disables that behavior.
+     *
+     * @param ignoreSdErrors ignore {@link ServiceDiscoverer} errors when evaluating a request failure.
+     * @return {@code this}.
+     */
+    public RetryingHttpRequesterFilterBuilder ignoreServiceDiscovererErrors(final boolean ignoreSdErrors) {
+        this.ignoreSdErrors = ignoreSdErrors;
+        return this;
+    }
+
+    /**
+     * Set the maximum number of allowed retry operations before giving up, applied as total max across both
+     * {@link #retryRequests(BiFunction)} and {@link #retryResponses(Function)}.
+     *
+     * @param maxRetries Maximum number of allowed retries before giving up
+     * @return {@code this}
+     */
+    public RetryingHttpRequesterFilterBuilder maxTotalRetries(final int maxRetries) {
+        if (maxRetries <= 0) {
+            throw new IllegalArgumentException("maxRetries: " + maxRetries + " (expected: >0)");
+        }
+        this.maxRetries = maxRetries;
+        return this;
+    }
+
+    /**
+     * Overrides the default criterion for determining which responses should be retried.
+     * <strong>It's important that this {@link Function} doesn't block to avoid performance impacts.</strong>
+     *
+     * @param mapper {@link Function} that checks whether a given {@link HttpResponseMetaData meta-data} should
+     * be retried, producing a {@link BackOffPolicy} in such cases.
+     * @return {@code this}
+     */
+    public RetryingHttpRequesterFilterBuilder retryResponses(
+            final Function<HttpResponseMetaData, BackOffPolicy> mapper) {
```

Comment:
Today, if users want to retry a response, they can map a status code into some exception type and handle it using a function that take `HttpRequestMetaData` and a `Throwable`. Do we have to add this new API?

Reply:
I think having everything in one place is more user friendly + results in one filter vs having more to do the mapping.

Reply:
We have 2 parallel discussions that are related:
1. `retryResponses`
2. `DefaultRequestRetryPolicyBuilder`
Let me continue here touching both of these points:

As you can see, it gets challenging to expose `retryResponses` and `retryRequests` on one filter. Not only because those are different. There are other related methods, like `maxTotalRetries`, which users may want to be different for request error or response status.

WDYT, if instead of two different retry points we still keep only one that handles `HttpRequestMetaData` and `Throwable`, but instead of `retryResponses(Function<HttpResponseMetaData, BackOffPolicy>)` we introduce something like `mapResponse(Function<HttpResponseMetaData, HttpResponseException>)`? `HttpResponseException` will have a reference to `HttpResponseMetaData`. That way users can configure mapping and keep using a single retry function (rename `retryRequests` to `retryOn/retryWhen`) to handle everything.

Reply:
Yup that's a good suggestion. Will do that now